src/ide/vscode/VscodeIDE.ts

```diff
@@ -1,51 +1,92 @@
-  workspace,
-  WorkspaceFolder,
-} from "vscode";
-import {
+import type * as vscode from "vscode";
+import { ExtensionContext, window, workspace, WorkspaceFolder } from "vscode";
+import type { TextDocumentChangeEvent } from "../../libs/common/ide/types/Events";
+import type {
   Disposable,
   IDE,
   RunMode,
 } from "../../libs/common/ide/types/ide.types";
 import VscodeClipboard from "./VscodeClipboard";
 import VscodeConfiguration from "./VscodeConfiguration";
+import { VscodeEditableTextEditorImpl } from "./VscodeEditableTextEditorImpl";
+import { vscodeOnDidChangeTextDocument } from "./VscodeEvents";
 import VscodeGlobalState from "./VscodeGlobalState";
 import VscodeMessages from "./VscodeMessages";
-
-const EXTENSION_MODE_MAP: Record<ExtensionMode, RunMode> = {
-  [ExtensionMode.Development]: "development",
-  [ExtensionMode.Production]: "production",
-  [ExtensionMode.Test]: "test",
-};
+import { vscodeRunMode } from "./VscodeRunMode";
+import { v4 as uuid } from "uuid";
 
 export default class VscodeIDE implements IDE {
   configuration: VscodeConfiguration;
   globalState: VscodeGlobalState;
   messages: VscodeMessages;
   clipboard: VscodeClipboard;
+  private editorMap;
 
   constructor(private extensionContext: ExtensionContext) {
     this.configuration = new VscodeConfiguration(this);
     this.globalState = new VscodeGlobalState(extensionContext);
     this.messages = new VscodeMessages();
     this.clipboard = new VscodeClipboard();
+    this.editorMap = new WeakMap<
+      vscode.TextEditor,
+      VscodeEditableTextEditorImpl
+    >();
   }
 
   get assetsRoot(): string {
     return this.extensionContext.extensionPath;
   }
 
   get runMode(): RunMode {
-    return EXTENSION_MODE_MAP[this.extensionContext.extensionMode];
+    return vscodeRunMode(this.extensionContext);
   }
 
   get workspaceFolders(): readonly WorkspaceFolder[] | undefined {
     return workspace.workspaceFolders;
   }
 
+  get activeTextEditor(): TextEditor | undefined {
+    return this.getActiveTextEditor();
+  }
+
+  get activeEditableTextEditor(): EditableTextEditor | undefined {
+    return this.getActiveTextEditor();
+  }
+
+  private getActiveTextEditor() {
+    return window.activeTextEditor != null
+      ? this.fromVscodeEditor(window.activeTextEditor)
+      : undefined;
+  }
+
+  get visibleTextEditors(): TextEditor[] {
+    return window.visibleTextEditors.map((e) => this.fromVscodeEditor(e));
+  }
+
+  public getEditableTextEditor(editor: TextEditor): EditableTextEditor {
+    return editor as VscodeEditableTextEditorImpl;
+  }
+
+  public onDidChangeTextDocument(
+    listener: (event: TextDocumentChangeEvent) => void,
+  ): Disposable {
+    return vscodeOnDidChangeTextDocument(listener);
+  }
+
+  public fromVscodeEditor(
+    editor: vscode.TextEditor,
+  ): VscodeEditableTextEditorImpl {
+    if (!this.editorMap.has(editor)) {
+      this.editorMap.set(
+        editor,
+        new VscodeEditableTextEditorImpl(uuid(), this, editor),
```

Comment:
This is now the only place where we instantiate editors.  There was a bug before where one editor could have two different ids, because `activeEditableTextEditor` was instantiating editors directly, ignoring the editor map

Reply:
That means that we could just remove `VscodeTextEditorImpl` because it's never actually used. Maybe we should keep the name `VscodeTextEditorImpl` as the single implementation and instead remove the verbose `VscodeEditableTextEditorImpl`?

Reply:
This of course means that any text editor could just be cast to an editable text editor anywhere in the code.

Reply:
> Maybe we should keep the name VscodeTextEditorImpl as the single implementation and instead remove the verbose VscodeEditableTextEditorImpl?

I don't feel strongly one way or another.  Would be tempted to just leave it for now, but if you want to merge them I'm not opposed; does get rid of the weird property override thing 🤷‍♂️

> This of course means that any text editor could just be cast to an editable text editor anywhere in the code.

I mean sure there's all kinds of horrible things the client code could do, but that's not really the fault of this design decision; the fact that there is one implementation is hidden behind the interface.  I think it's actually a good idea to allow one implementation to appear in different places via different interfaces in order to hide irrelevant details from the client

Reply:
Ok will go ahead and do this.

Reply:
btw why is passing uuid in better then creating it in the editor constructor?

Reply:
i don't feel super strongly bout that one, just felt like we wanted one centralised place managing the id's.  then eg we could switch from uuid to sequential, could easily inject id's during a test, etc

Reply:
Ok I will leave the id as a parameter.